google-cloud-storage/src/main/java/com/google/cloud/storage/BlobReadChannelV2.java

```diff
@@ -0,0 +1,295 @@
+          bufferHandle = null;
+        }
+        lazyReadChannel = null;
+      }
+    }
+  }
+
+  private LazyReadChannel<StorageObject> internalGetLazyChannel() {
+    if (lazyReadChannel == null) {
+      lazyReadChannel = newLazyReadChannel();
+    }
+    return lazyReadChannel;
+  }
+
+  private LazyReadChannel<StorageObject> newLazyReadChannel() {
+    return new LazyReadChannel<>(
+        () -> {
+          if (bufferHandle == null) {
+            bufferHandle = BufferHandle.allocate(chunkSize);
+          }
+          return ResumableMedia.http()
+              .read()
+              .byteChannel(clientStuff)
+              .setCallback(this::setResolvedObject)
+              .buffered(bufferHandle)
+              .setApiaryReadRequest(getApiaryReadRequest())
+              .build();
+        });
+  }
+
+  private void setResolvedObject(StorageObject resolvedObject) {
+    this.resolvedObject = resolvedObject;
+  }
+
+  private ApiaryReadRequest getApiaryReadRequest() {
+    // TODO: test what happens if you try to read from the last byte of an object
+    StorageObject object = resolvedObject != null ? resolvedObject : storageObject;
+    return new ApiaryReadRequest(object, opts, byteRangeSpec);
+  }
+
+  static class BlobReadChannelV2State implements RestorableState<ReadChannel>, Serializable {
+
+    private static final long serialVersionUID = -7595661593080505431L;
+
+    private final ApiaryReadRequest request;
+    private final ClientStuff clientStuff;
+
+    private final Integer chunkSize;
+
+    private BlobReadChannelV2State(
+        ApiaryReadRequest request, ClientStuff clientStuff, Integer chunkSize) {
+      this.request = request;
+      this.clientStuff = clientStuff;
+      this.chunkSize = chunkSize;
+    }
+
+    @Override
+    public ReadChannel restore() {
+      BlobReadChannelV2 channel =
+          new BlobReadChannelV2(request.getObject(), request.getOptions(), clientStuff);
+      channel.setByteRangeSpec(request.getByteRangeSpec());
+      if (chunkSize != null) {
+        channel.setChunkSize(chunkSize);
+      }
+      return channel;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (!(o instanceof BlobReadChannelV2State)) {
+        return false;
+      }
+      BlobReadChannelV2State that = (BlobReadChannelV2State) o;
+      return Objects.equals(request, that.request)
+          && Objects.equals(clientStuff, that.clientStuff)
+          && Objects.equals(chunkSize, that.chunkSize);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(request, clientStuff, chunkSize);
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("\nrequest", request)
+          .add("\nclientStuff", clientStuff)
+          .add("\nchunkSize", chunkSize)
+          .toString();
+    }
+  }
+
+  static final class ClientStuff implements Serializable {
```

Comment:
Any opinions on what we call this? The name needs to stay stable due to it being serialized out so we probably want something better than `ClientStuff`.

Reply:
Ended up not needing this class for capture, so the name here is less important. I've moved to using BlobReadChannelContext for the time being.